File: pyfenstein3d/engine/vector2d.py
"""contem classe Vector2d"""

import math

class Vector2d():
    """permite realizar calculos vetoriais"""
    def __init__(self, x: float = 0, y: float = 0):
        self.__x = x
        self.__y = y
        self.__ang = (math.atan2(self.__y, self.__x) + math.pi * 2) % (math.pi * 2)
        self.__mag = math.sqrt(math.pow(self.__x, 2) + math.pow(self.__y, 2))

    def get_x(self) -> float:
        """obtem x"""
        return self.__x
    x: float = property(get_x)

    def get_y(self) -> float:
        """obtem y"""
        return self.__y
    y: float = property(get_y)

    def get_mag(self) -> float:
        """obtem magnitude"""
        return self.__mag
    mag: float = property(get_mag)

    def get_ang(self) -> float:
        """obtem angulo"""
        return self.__ang
    ang = property(get_ang)

    def copy(self):
        """cria copia"""
        return Vector2d(self.x, self.y)

    def __add__(self, other):
        return Vector2d(self.x + other.x, self.y + other.y)

    def __sub__(self, other):
        return Vector2d(self.x - other.x, self.y - other.y)

    def __mul__(self, other: float):
        return Vector2d(self.x * other, self.y * other)

    def __truediv__(self, other: float):
        return Vector2d(self.x / other, self.y / other)

    def __floordiv__(self, other: float):
        return Vector2d(self.x // other, self.y // other)

    def __pow__(self, other: float):
        cos = math.cos(other)
        sin = math.sin(other)
        v_x = self.x
        v_y = self.y
        return Vector2d(v_x * cos - v_y * sin, v_x * sin + v_y * cos)

    def __mod__(self, other: float):
        mag = self.mag
        if mag == 0:
            return self
        return Vector2d((self.__x / mag) * other, (self.__y / mag) * other)

    @staticmethod
    def create_with_ang(rad: float):
        return Vector2d(math.cos(rad), math.sin(rad))

File: engine/vector2d.go
package engine

import "math"

// Vector2d represents a 2D vector with precomputed angle and magnitude.
type Vector2d struct {
	x, y, ang, mag float64
}

// NewVector2d creates a Vector2d from x and y components.
func NewVector2d(x, y float64) *Vector2d {
	ang := math.Mod(math.Atan2(y, x)+math.Pi*2, math.Pi*2)
	mag := math.Sqrt(x*x + y*y)
	return &Vector2d{x: x, y: y, ang: ang, mag: mag}
}

// CreateWithAng creates a unit Vector2d from an angle in radians.
func CreateWithAng(rad float64) *Vector2d {
	return NewVector2d(math.Cos(rad), math.Sin(rad))
}

func (v *Vector2d) X() float64   { return v.x }
func (v *Vector2d) Y() float64   { return v.y }
func (v *Vector2d) Ang() float64 { return v.ang }
func (v *Vector2d) Mag() float64 { return v.mag }

// Copy returns a new Vector2d with the same components.
func (v *Vector2d) Copy() *Vector2d { return NewVector2d(v.x, v.y) }

// Add returns v + other.
func (v *Vector2d) Add(other *Vector2d) *Vector2d {
	return NewVector2d(v.x+other.x, v.y+other.y)
}

// Sub returns v - other.
func (v *Vector2d) Sub(other *Vector2d) *Vector2d {
	return NewVector2d(v.x-other.x, v.y-other.y)
}

// Mul returns v * scalar.
func (v *Vector2d) Mul(scalar float64) *Vector2d {
	return NewVector2d(v.x*scalar, v.y*scalar)
}

// Div returns v / scalar.
func (v *Vector2d) Div(scalar float64) *Vector2d {
	return NewVector2d(v.x/scalar, v.y/scalar)
}

// FloorDiv returns floor(v / scalar) component-wise.
func (v *Vector2d) FloorDiv(scalar float64) *Vector2d {
	return NewVector2d(math.Floor(v.x/scalar), math.Floor(v.y/scalar))
}

// Rot rotates v by rad radians and returns the result (equivalent to ** in Python).
func (v *Vector2d) Rot(rad float64) *Vector2d {
	cos := math.Cos(rad)
	sin := math.Sin(rad)
	return NewVector2d(v.x*cos-v.y*sin, v.x*sin+v.y*cos)
}

// Norm returns a vector in the same direction as v with the given magnitude (equivalent to % in Python).
func (v *Vector2d) Norm(mag float64) *Vector2d {
	if v.mag == 0 {
		return v
	}
	return NewVector2d((v.x/v.mag)*mag, (v.y/v.mag)*mag)
}
